File: elixir/secret-handshake/secret_handshake.exs
defmodule SecretHandshake do
  import Enum
  @doc """
  Determine the actions of a secret handshake based on the binary
  representation of the given `code`.

  If the following bits are set, include the corresponding action in your list
  of commands, in order from lowest to highest.

  1 = wink
  10 = double blink
  100 = close your eyes
  1000 = jump

  10000 = Reverse the order of the operations in the secret handshake
  """
  @spec commands(code :: integer) :: list(String.t())
  def commands(code) do
    process_command = fn str, cmds ->
      add = &([&1] ++ cmds)
      case str do
        "1" -> add.("wink")
        "10" -> add.("double blink")
        "100" -> add.("close your eyes")
        "1000" -> add.("jump")
        "10000" -> List.reverse cmdsexer
      end
    end
    split_into_commands = fn(str) ->
      hd = &Kernel.hd/1
      Regex.scan(~r/(10*)/,str) |> map(hd)
    end
    code
      |> Integer.digits(2)
      |> map(&to_charlist/1)
      |> join
      |> split_into_commands.()
      |> reduce([], process_command)
      |> reverse
  end
end

File: typescript/secret-handshake/secret-handshake.ts
enum Command {
    Wink = 1,
    DoubleBlink,
    CloseEyes,
    Jump,
    Reverse,
}



const CommandStrings: {
    [key: number]: string
} = {
    [Command.Wink]: "wink",
    [Command.DoubleBlink]: "double blink",
    [Command.CloseEyes]: "close your eyes",
    [Command.Jump]: "jump"
}


export default class Handshake {
    constructor(private integer: number) { }
    private binary(): string {
        return new Number(this.integer).toString(2)
    }
    commands(): string[] {
        const binaryString = this.binary()
        const pad5DigitsWithZeros = (s: string): string => {
            if (s.length == 5)
                return s
            else return pad5DigitsWithZeros("0" + s)
        }
        const paddedBinaryString = pad5DigitsWithZeros(binaryString);
        const commandForIndex = (idx: number) => 5 - idx
        const reducer = (commands: Command[], element: string, idx: number): Command[] => {
            if (element == "1")
                return commands.concat([commandForIndex(idx)])
            return commands;
        }
        let commands: Command[] = paddedBinaryString.split("")
            .reduce(reducer, [])
        if (commands[0] && commands[0] == Command.Reverse) {
            const [, ...commandsRest] = commands;
            commands = commandsRest.reverse()
        }
        return commands.reverse().map((c) => CommandStrings[c])
    }
}